scripts/lazy-load-images.js

```diff
@@ -1,20 +1,56 @@
-const inView = require('in-view')
+import { throttle } from 'lodash'
+
+function inViewport (element) {
+  const { top, right, bottom, left } = element.getBoundingClientRect()
+  return bottom > 0
+      && window.innerWidth - left > 0
+      && window.innerHeight - top > 0
+      && right > 0
+}
 
 module.exports = function () {
-  inView('img[data-src]')
-    .on('enter', img => {
-      // abort if this element has previously entered the viewport
-      if (!img.dataset.src) return
-
-      img.setAttribute('src', img.dataset.src)
-
-      // remove attribute so CSS fade-in can happen
-      img.onload = () => {
-        img.removeAttribute('data-src')
-        
-        // remove colored circular placeholder element
-        const placeholder = img.parentElement.querySelector('.placeholder')
-        if (placeholder) placeholder.classList.add('faded-out')
-      }
-    })
-}
\ No newline at end of file
+  ['scroll', 'resize', 'load'].forEach(event =>
+    window.addEventListener(event, updateImages)
+  )
+
+  // add all images that are part of the DOM tree
+  // NB: |display: none| elements need manual adding
+  document.querySelectorAll('img[data-src]').forEach(addImage)
+}
+
+let images = []
+
+let addImage = module.exports.addImage = function (image) {
+  // abort if this element has previously entered the viewport
+  if (!image.dataset.src) return
+
+  images.push(image)
+  updateImages()
+}
+
+function handle (image) {
+  // abort if this element has entered the viewport after added to the list
+  if (!image.dataset.src) return
+
+  image.setAttribute('src', image.dataset.src)
+
+  // remove attribute so CSS fade-in can happen
+  image.onload = () => {
+    image.removeAttribute('data-src')
+
+    // remove colored circular placeholder element
+    const placeholder = image.parentElement.querySelector('.placeholder')
+    if (placeholder) placeholder.classList.add('faded-out')
+  }
+}
+
+// TODO: Read/write lock on |images|?
```

Comment:
Can you clarify what this TODO mean?

Reply:
Since I applied a filter over the list, I am thinking if there will be cases when `addImage` is called (adding another element into the list) while the loop is taking place.

Do we want to have:

```
let copy = images;
images = []
// test & filter elements in copy
// move elements from images (probably not empty because `addImage` is called)
// images = copy
```